src/perform_step/composite_perform_step.jl

```diff
@@ -1,38 +1,61 @@
-#=
-
-Maybe do generated functions to reduce dispatch times?
-
-f(x) = x
-g(x,i) = f(x[i])
-g{i}(x,::Type{Val{i}}) = f(x[i])
-@generated function gg(tup::Tuple, num)
-                  N = length(tup.parameters)
-                  :(@nif $(N+1) i->(i == num) i->(f(tup[i])) i->error("unreachable"))
-              end
-h(i) = g((1,1.0,"foo"), i)
-h2{i}(::Type{Val{i}}) = g((1,1.0,"foo"), Val{i})
-h3(i) = gg((1,1.0,"foo"), i)
-@benchmark h(1)
-mean time:        31.822 ns (0.00% GC)
-@benchmark h2(Val{1})
-mean time:        1.585 ns (0.00% GC)
-@benchmark h3(1)
-mean time:        6.423 ns (0.00% GC)
-
-@generated function foo(tup::Tuple, num)
-  N = length(tup.parameters)
-  :(@nif $(N+1) i->(i == num) i->(tup[i]) i->error("unreachable"))
+function init_ith_default_cache(cache::DefaultCache, algs, i)
+    if i == 1
```

Comment:
It can be reduced to
```julia
if 1 <= i <= 6
    s = Symbol("cache$i")
    if !isdefined(cache, s)
        setproperty!(cache, s, alg_cache(algs[i], cache.args...))
    end
end
```

Reply:
this code is very specifically written to be type stable, which this suggestion wouldn't be.